src/python/pants/backend/python/check/pip_audit/rules.py

```diff
@@ -0,0 +1,169 @@
+from pants.base.build_root import BuildRoot
+from pants.core.goals.check import CheckRequest, CheckResult, CheckResults
+from pants.engine.collection import Collection
+from pants.engine.fs import CreateDigest, Digest, FileContent
+from pants.engine.process import FallibleProcessResult, ProcessCacheScope
+from pants.engine.rules import Get, MultiGet, collect_rules, rule
+from pants.engine.target import FieldSet, Target
+from pants.engine.unions import UnionRule
+from pants.util.logging import LogLevel
+from pants.util.ordered_set import FrozenOrderedSet, OrderedSet
+from pants.util.strutil import pluralize
+
+
+@dataclass(frozen=True)
+class PipAuditFieldSet(FieldSet):
+    required_fields = (
+        PythonRequirementsField,
+        PythonRequirementResolveField,
+    )
+
+    resolve: PythonRequirementResolveField
+    sources: PythonRequirementsField
+
+    @classmethod
+    def opt_out(cls, tgt: Target) -> bool:
+        return tgt.get(SkipPipAuditField).value
+
+
+@dataclass(frozen=True)
+class PipAuditPartition:
+    description: str
+    field_sets: FrozenOrderedSet[PipAuditFieldSet]
+
+
+class PipAuditPartitions(Collection[PipAuditPartition]):
+    pass
+
+
+class PipAuditRequest(CheckRequest):
+    field_set_type = PipAuditFieldSet
+    name = PipAudit.options_scope
+
+
+@rule
+async def pip_audit_partition(
+    partition: PipAuditPartition,
+    pip_audit: PipAudit,
+    build_root: BuildRoot,
+) -> CheckResult:
+    requirements_pex_get = Get(
+        VenvPex,
+        RequirementsPexRequest(
+            (fs.address for fs in partition.field_sets),
+        ),
+    )
+
+    pip_audit_pex_get = Get(
+        VenvPex,
+        PexRequest,
+        pip_audit.to_pex_request(),
+    )
+
+    requirements_pex, pip_audit_pex = await MultiGet(requirements_pex_get, pip_audit_pex_get)
+    requirements_pex_info = await Get(PexResolveInfo, VenvPex, requirements_pex)
+    requirements_str = "\n".join(
+        "{}=={}".format(dist_info.project_name, dist_info.version)
+        for dist_info in requirements_pex_info
+    )
+    requirements_digest = await Get(
+        Digest,
+        CreateDigest([FileContent("requirements.txt", requirements_str.encode())]),
+    )
+
+    cache_dir = f".cache/pip_audit_cache/{sha256(build_root.path.encode()).hexdigest()}"
+    result = await Get(
+        FallibleProcessResult,
+        VenvPexProcess(
+            pip_audit_pex,
+            argv=(
+                "--cache-dir",
+                cache_dir,
+                "--no-deps",
+                "--progress-spinner",
+                "off",
+                "--requirement",
+                "requirements.txt",
+                *pip_audit.args,
+            ),
+            description=f"Run pip-audit on {pluralize(len(requirements_pex_info), 'requirement')}.",
+            level=LogLevel.DEBUG,
+            input_digest=requirements_digest,
+            cache_scope=ProcessCacheScope.PER_SESSION,
```

Comment:
I added this because caching the audit makes little sense since a new vulnerability could be published between runs. Thinking about it now however, surely some of the `@rules` need something similar added to them or Pants won't even get here with a warm cache?

Reply:
If any child of a `@rule` is uncacheable like this, then all of the parents are uncacheable. So just marking a process as per-session is sufficient.

Reply:
Good to know. Thanks!